src/drivers/imu/bosch/bmi088/BMI088_Accelerometer.cpp

```diff
@@ -0,0 +1,601 @@
+			r.set_bits = fifo_watermark_threshold & 0x00FF;
+			r.clear_bits = ~r.set_bits;
+
+		} else if (r.reg == Register::FIFO_WTM_1) {
+			// fifo_water_mark[12:8]
+			r.set_bits = (fifo_watermark_threshold & 0x0700) >> 8;
+			r.clear_bits = ~r.set_bits;
+		}
+	}
+}
+
+bool BMI088_Accelerometer::Configure()
+{
+	// first set and clear all configured register bits
+	for (const auto &reg_cfg : _register_cfg) {
+		RegisterSetAndClearBits(reg_cfg.reg, reg_cfg.set_bits, reg_cfg.clear_bits);
+	}
+
+	// now check that all are configured
+	bool success = true;
+
+	for (const auto &reg_cfg : _register_cfg) {
+		if (!RegisterCheck(reg_cfg)) {
+			success = false;
+		}
+	}
+
+	ConfigureAccel();
+
+	return success;
+}
+
+int BMI088_Accelerometer::DataReadyInterruptCallback(int irq, void *context, void *arg)
+{
+	static_cast<BMI088_Accelerometer *>(arg)->DataReady();
+	return 0;
+}
+
+void BMI088_Accelerometer::DataReady()
+{
+	uint8_t expected = 0;
+
+	if (_drdy_fifo_read_samples.compare_exchange(&expected, _fifo_accel_samples)) {
+		ScheduleNow();
+	}
+}
+
+bool BMI088_Accelerometer::DataReadyInterruptConfigure()
+{
+	if (_drdy_gpio == 0) {
+		return false;
+	}
+
+	// Setup data ready on falling edge
+	return px4_arch_gpiosetevent(_drdy_gpio, false, true, true, &DataReadyInterruptCallback, this) == 0;
+}
+
+bool BMI088_Accelerometer::DataReadyInterruptDisable()
+{
+	if (_drdy_gpio == 0) {
+		return false;
+	}
+
+	return px4_arch_gpiosetevent(_drdy_gpio, false, false, false, nullptr, nullptr) == 0;
+}
+
+bool BMI088_Accelerometer::RegisterCheck(const register_config_t &reg_cfg)
+{
+	bool success = true;
+
+	const uint8_t reg_value = RegisterRead(reg_cfg.reg);
+
+	if (reg_cfg.set_bits && ((reg_value & reg_cfg.set_bits) != reg_cfg.set_bits)) {
+		PX4_DEBUG("0x%02hhX: 0x%02hhX (0x%02hhX not set)", (uint8_t)reg_cfg.reg, reg_value, reg_cfg.set_bits);
+		success = false;
+	}
+
+	if (reg_cfg.clear_bits && ((reg_value & reg_cfg.clear_bits) != 0)) {
+		PX4_DEBUG("0x%02hhX: 0x%02hhX (0x%02hhX not cleared)", (uint8_t)reg_cfg.reg, reg_value, reg_cfg.clear_bits);
+		success = false;
+	}
+
+	return success;
+}
+
+uint8_t BMI088_Accelerometer::RegisterRead(Register reg)
+{
+	// 6.1.2 SPI interface of accelerometer part
+	//
+	// In case of read operations of the accelerometer part, the requested data
+	// is not sent immediately, but instead first a dummy byte is sent, and
+	// after this dummy byte the actual requested register content is transmitted.
+	uint8_t cmd[3] {};
+	cmd[0] = static_cast<uint8_t>(reg) | DIR_READ;
+	// cmd[1] dummy byte
+	transfer(cmd, cmd, sizeof(cmd));
+	return cmd[2];
+}
+
+void BMI088_Accelerometer::RegisterWrite(Register reg, uint8_t value)
+{
+	uint8_t cmd[2] {(uint8_t)reg, value};
+	transfer(cmd, cmd, sizeof(cmd));
+}
+
+void BMI088_Accelerometer::RegisterSetAndClearBits(Register reg, uint8_t setbits, uint8_t clearbits)
+{
+	const uint8_t orig_val = RegisterRead(reg);
+
+	uint8_t val = (orig_val & ~clearbits) | setbits;
```

Comment:
Since this is now clear and then set (which does make more sense), I'd swap the order in the argument and struct definition.
(what would also be nice and more robust is to have masks, so all bits are cleared in a multi-bit field - what you did manually now in `ACC_CONF`).

Reply:
I'm just debating doing it now for the BMI088, or keeping it consistent with the others and doing it everywhere in one pass with shared register methods on the SPI interface.

I agree about the masks, the clear and set bits table was a bit overly simplistic, but it was easy to get away with for the initial Invensense IMUs where nearly everything is first reset to 0. I was looking at options for actually representing the range of bits of a register, but that might be a rabbit hole with diminishing returns.